### PR TITLE
UCS/CONFIG: Filtered out sections from config by platform info.

### PR DESCRIPTION
**Pull request by rakhmets**
## What
Added "Product name" to the output of `ucx_info -s`.
Added ability to skip sections of config file, if it is filtered out by "CPU vendor", "CPU model", or "Product name".

### CODE REVIEW

File: src/ucs/sys/sys.c [hunk 1]
```diff
@@ -1648,3 +1649,21 @@ ucs_status_t ucs_sys_read_sysfs_file(const char *dev_name,
...
+    static int initialized        = 0;
+    ssize_t nread;
+
+    if (!initialized) {
+        initialized = 1;
+        nread       = ucs_read_file_str(product_name, sizeof(product_name), 1,
+                                        UCS_DMI_PRODUCT_NAME_FILE);
+        if (nread >= 0) {
+            ucs_strtrim(product_name);
+        }
```
Comment (tvegas1) [inline src/ucs/sys/sys.c:1665]: else do we dump something like `N/A` or empty string?

File: src/ucs/config/parser.c [hunk 2]
```diff
@@ -1512,16 +1525,62 @@ static char *ucs_config_get_value_from_config_file(const char *name)
     return kh_val(&ucs_config_file_vars, iter);
 }
 
+static int ucs_config_parse_check_filter(const char *name, const char *value)
+{
+    typedef struct filter {
```
Comment (tvegas1) [inline src/ucs/config/parser.c:1530]: `filter` can be removed

File: src/ucs/config/parser.c [hunk 3]
```diff
@@ -1512,16 +1525,62 @@ static char *ucs_config_get_value_from_config_file(const char *name)
...
+        char *name;
+        const char *(*value_f)();
+    } filter_t;
+
+    static filter_t filters[3] = {{UCS_CPU_VENDOR_LABEL, ucs_cpu_vendor_name},
```
Comment (tvegas1) [inline src/ucs/config/parser.c:1535]: directly []?

File: src/ucs/config/parser.c [hunk 4]
```diff
@@ -1554,8 +1614,12 @@ void ucs_config_parse_config_file(const char *dir_path, const char *file_name,
         return;
     }
 
+    parse_arg.override = override;
```
Comment (tvegas1) [inline src/ucs/config/parser.c:1617]: maybe to replace the three line:
`ucs_config_parse_arg_t parse_arg = { .override = override };`

**Review by tvegas1 — COMMENTED** [review]
cosmetic comments

File: src/ucs/config/parser.c [hunk 5]
```diff
@@ -1512,16 +1525,62 @@ static char *ucs_config_get_value_from_config_file(const char *name)
...
+        const char *(*value_f)();
+    } filter_t;
+
+    static filter_t filters[3] = {{UCS_CPU_VENDOR_LABEL, ucs_cpu_vendor_name},
+                                  {UCS_CPU_MODEL_LABEL, ucs_cpu_model_name},
+                                  {UCS_SYS_DMI_PRODUCT_NAME_LABEL,
+                                   ucs_sys_dmi_product_name}};
+    int i;
+
+    for (i = 0; i < ucs_static_array_size(filters); ++i) {
```
Comment (tvegas1) [inline src/ucs/config/parser.c:1541]: `ucs_carray_for_each()` instead?

File: src/ucs/config/parser.c (hunk 4 above)
Comment (rakhmets) [reply src/ucs/config/parser.c]: Replaced by initialization during declaration.

File: src/ucs/config/parser.c (hunk 3 above)
Comment (rakhmets) [reply src/ucs/config/parser.c]: Removed.

File: src/ucs/config/parser.c (hunk 2 above)
Comment (rakhmets) [reply src/ucs/config/parser.c]: Removed.

File: src/ucs/sys/sys.c (hunk 1 above)
Comment (rakhmets) [reply src/ucs/sys/sys.c:1666]: Dump `N/A` in case of failure.

File: src/ucs/sys/sys.c [hunk 6]
```diff
@@ -1648,3 +1649,21 @@ ucs_status_t ucs_sys_read_sysfs_file(const char *dev_name,
...
     return UCS_OK;
 }
+
+const char *ucs_sys_dmi_product_name()
+{
+    static char product_name[128] = {0};
+    static int initialized        = 0;
+    ssize_t nread;
+
+    if (!initialized) {
```
[the remark this answers is not in the document]
Comment (rakhmets) [reply src/ucs/sys/sys.c]: Added `UCS_INIT_ONCE()`.

**Review by tvegas1 — DISMISSED** [review]

File: test/gtest/ucs/ucx_test.conf [hunk 7]
```diff
@@ -1,3 +1,26 @@
...
+CPU vendor=*
+CPU model=*
+Product name=*
 UCX_TEMP=10,front:15
+
+[section to skip 1]
+CPU vendor=Skip section
+CPU model=*
+Product name=*
+UCX_PRICE=300
```
Comment (brminich) [inline test/gtest/ucs/ucx_test.conf:14]: can you pls provide some example how this skip feature will be used?

Comment (rakhmets) [reply test/gtest/ucs/ucx_test.conf:14]: Section is used, if filtering patterns are matched.
E.g. we have the following config
```
[section1]
CPU model=*model_1*
UCX_VALUE=value_1

[section2]
Product name=*product_2*
UCX_VALUE=value_2
```
, then:
```
$ ucx_info -s
...
# CPU vendor: vendor_1
# CPU model: model_1 update
# Product name: product_1
...
$ ucx_info -c | grep UCX_VALUE
UCX_VALUE=value_1
```
, or:
```
$ ucx_info -s
...
# CPU vendor: vendor_2
# CPU model: model_2
# Product name: product_2
...
$ ucx_info -c | grep UCX_VALUE
UCX_VALUE=value_2
```

File: src/ucs/config/parser.c [hunk 8]
```diff
@@ -70,6 +70,19 @@ const char *ucs_async_mode_names[] = {
 
 UCS_CONFIG_DEFINE_ARRAY(string, sizeof(char*), UCS_CONFIG_TYPE_STRING);
 
+
+typedef struct ucs_config_parse_section {
+    char name[64];
```
Comment (brminich) [inline src/ucs/config/parser.c:75]: where is this section name used?

Comment (rakhmets) [reply src/ucs/config/parser.c:75]: In the first condition in `ucs_config_parse_skip_section`.

**Review by brminich — DISMISSED** [review]

File: src/ucs/config/parser.c [hunk 9]
```diff
@@ -1512,16 +1525,63 @@ static char *ucs_config_get_value_from_config_file(const char *name)
...
+                                 {UCS_CPU_MODEL_LABEL, ucs_cpu_model_name},
+                                 {UCS_SYS_DMI_PRODUCT_NAME_LABEL,
+                                  ucs_sys_dmi_product_name}};
+
+    filter_t *filter;
+
+    ucs_carray_for_each(filter, filters, ucs_static_array_size(filters)) {
+        if ((strcmp(name, filter->name) == 0) &&
+            (fnmatch(value, filter->value_f(), FNM_CASEFOLD) != 0)) {
+            return 1;
```
Comment (yosefe) [inline src/ucs/config/parser.c:1545]: pls add comment to explain

File: src/ucs/config/parser.c [hunk 10]
```diff
@@ -1512,16 +1525,63 @@ static char *ucs_config_get_value_from_config_file(const char *name)
...
+    return 0;
+}
+
+static int
+ucs_config_parse_skip_section(ucs_config_parse_section_t *section_info,
+                              const char *section, const char *name,
+                              const char *value)
+{
+    if (strcmp(section, section_info->name) != 0) {
+        ucs_strncpy_zero(section_info->name, section,
```
Comment (yosefe) [inline src/ucs/config/parser.c:1558]: pls add comment to explain

File: src/ucs/config/parser.c [hunk 11]
```diff
@@ -1512,16 +1525,63 @@ static char *ucs_config_get_value_from_config_file(const char *name)
...
                                              const char *name,
                                              const char *value)
 {
-    khiter_t iter = kh_get(ucs_config_map, &ucs_config_file_vars, name);
-    int override  = *(int*)arg;
+    ucs_config_parse_arg_t *parse_arg = (ucs_config_parse_arg_t*)arg;
+    khiter_t iter;
     int result;
 
+    if (ucs_config_parse_skip_section(&parse_arg->section_info, section, name,
```
Comment (yosefe) [inline src/ucs/config/parser.c:1577]: optional: IMO the code could be a little simpler if ucs_config_parse_skip_section returned void, and we checked parse_arg->section_info.skip directly in ucs_config_parse_config_file_line.

File: src/ucs/sys/sys.c [hunk 12]
```diff
@@ -1648,3 +1649,22 @@ ucs_status_t ucs_sys_read_sysfs_file(const char *dev_name,
...
+    static char product_name[128]    = {0};
+    ssize_t nread;
+
+    UCS_INIT_ONCE(&init_once) {
+        nread = ucs_read_file_str(product_name, sizeof(product_name), 1,
+                                  UCS_DMI_PRODUCT_NAME_FILE);
+        if (nread >= 0) {
+            ucs_strtrim(product_name);
+        } else {
+            ucs_strncpy_zero(product_name, "N/A", sizeof(product_name));
```
[the remark this answers is not in the document]
Comment (rakhmets) [reply src/ucs/sys/sys.c]: Added constant to use in these methods.

File: src/ucs/sys/sys.h [hunk 13]
```diff
@@ -67,6 +67,9 @@ typedef cpuset_t ucs_sys_cpuset_t;
 #define UCS_SYS_FS_CPUS_PATH   UCS_SYS_FS_SYSTEM_PATH "/cpu"
 
 
+#define UCS_SYS_DMI_PRODUCT_NAME_LABEL "Product name"
```
[the remark this answers is not in the document]
Comment (rakhmets) [reply src/ucs/sys/sys.h:70]: But I use it in `sys_info.c`. Then the output of `ucx_info -s` can be used to create filters for section in a configuration file.

File: src/ucs/config/parser.c [hunk 14]
```diff
@@ -1512,16 +1525,63 @@ static char *ucs_config_get_value_from_config_file(const char *name)
     return kh_val(&ucs_config_file_vars, iter);
 }
 
+static int ucs_config_parse_check_filter(const char *name, const char *value)
+{
+    typedef struct {
+        char *name;
```
[the remark this answers is not in the document]
Comment (rakhmets) [reply src/ucs/config/parser.c]: Added.

File: src/ucs/config/parser.c [hunk 15]
```diff
@@ -1512,16 +1525,63 @@ static char *ucs_config_get_value_from_config_file(const char *name)
...
+        const char *(*value_f)();
+    } filter_t;
+
+    static filter_t filters[] = {{UCS_CPU_VENDOR_LABEL, ucs_cpu_vendor_name},
+                                 {UCS_CPU_MODEL_LABEL, ucs_cpu_model_name},
+                                 {UCS_SYS_DMI_PRODUCT_NAME_LABEL,
+                                  ucs_sys_dmi_product_name}};
+
```
[the remark this answers is not in the document]
Comment (rakhmets) [reply src/ucs/config/parser.c]: Removed.

File: test/gtest/ucs/ucx_test.conf [hunk 16]
```diff
@@ -1,3 +1,26 @@
...
 UCX_BRAND=Mazda
+
+[section to use]
+CPU vendor=*
+CPU model=*
+Product name=*
 UCX_TEMP=10,front:15
+
+[section to skip 1]
+CPU vendor=Skip section
```
[the remark this answers is not in the document]
Comment (rakhmets) [reply test/gtest/ucs/ucx_test.conf]: Updated.

File: test/gtest/ucs/ucx_test.conf [hunk 17]
```diff
@@ -1,3 +1,26 @@
 UCX_PRICE=200
 UCX_BRAND=Mazda
+
+[section to use]
+CPU vendor=*
+CPU model=*
+Product name=*
```
[the remark this answers is not in the document]
Comment (rakhmets) [reply test/gtest/ucs/ucx_test.conf]: Removed.

File: src/ucs/config/parser.c (hunk 10 above)
Comment (rakhmets) [reply src/ucs/config/parser.c:1563]: Added a comment.

File: src/ucs/config/parser.c (hunk 11 above)
Comment (rakhmets) [reply src/ucs/config/parser.c]: Updated: `int ucs_config_parse_skip_section` -> `void ucs_config_parse_set_section_info`.

File: test/gtest/ucs/ucx_test.conf [hunk 18]
```diff
@@ -2,25 +2,22 @@ UCX_PRICE=200
...
-CPU vendor=Skip section
+CPU vendor=Invalid CPU vendor name
 CPU model=*
 Product name=*
 UCX_PRICE=300
 
 [section to skip 2]
 CPU vendor=*
-CPU model=Skip section
+CPU model=Invalid CPU vendor name
```
Comment (yosefe) [inline test/gtest/ucs/ucx_test.conf:15]: Invalid CPU model name

File: src/ucs/config/parser.c [hunk 19]
```diff
@@ -1512,16 +1525,69 @@ static char *ucs_config_get_value_from_config_file(const char *name)
...
+static void
+ucs_config_parse_set_section_info(ucs_config_parse_section_t *section_info,
+                                  const char *section, const char *name,
+                                  const char *value)
+{
+    if (strcmp(section, section_info->name) != 0) {
+        /* A new section has started. Update section information. */
+        ucs_strncpy_zero(section_info->name, section,
+                         sizeof(section_info->name));
+        section_info->skip = 0;
```
Comment (yosefe) [inline src/ucs/config/parser.c:1565]: can we remove this line? section_info->skip is set anyway in line 1571

File: src/ucs/sys/string.h [hunk 20]
```diff
@@ -37,6 +37,9 @@ BEGIN_C_DECLS
 #define UCS_HEXUNITS_AUTO   ((uint16_t)-2)
 
 
+#define UCS_VALUE_UNKOWN_STR "unknown"
```
Comment (yosefe) [inline src/ucs/sys/string.h:40]: UCS_VALUE_UNKOWN_STR -> UCS_VALUE_UNK**N**OWN_STR

File: test/gtest/ucs/ucx_test.conf (hunk 18 above)
Comment (rakhmets) [reply test/gtest/ucs/ucx_test.conf]: Fixed.

File: src/ucs/config/parser.c (hunk 19 above)
Comment (rakhmets) [reply src/ucs/config/parser.c]: Removed.

File: src/ucs/sys/string.h (hunk 20 above)
Comment (rakhmets) [reply src/ucs/sys/string.h]: Fixed.

**Review by yosefe — APPROVED** [review]

**Review by tvegas1 — APPROVED** [review]

**Pull request merged**